climt/_core/climt_components.py

```diff
@@ -363,6 +363,9 @@ def outputs(self):
     def diagnostics(self):
         return tuple(self._climt_diagnostics.keys())
 
+    def prognostic_version(self):
```

Comment:
This method shouldn't exist. It's no easier to write implicit.prognostic_version() than it is to write ClimtTimeDifferenced(implicit) (or whatever we do down the line). We should try to avoid having two equally valid ways to do the same thing (it's a Python idiom). It makes it more confusing to learn the right way to do things.

Reply:
I prefer this way of accessing the wrappers. This allows, in an interactive setup, the user to just press tab and see what "behaviours" are available. If the issue is that ClimtTimeDifferenced is part of the public API, I would have no issues removing it. I just don't want people to need to remember which wrapper to import to get something done.

Reply:
If it's really a hassle to remember which wrapper to import to get something done, then Sympl should not expose e.g. UpdateFrequencyWrapper as objects, and instead should expose methods like the ones you've used. However, we *have* to expose those objects if we have these methods, since they're the type that this method returns, and it's bad form to return a private type.

It should be easy to see what wrappers are available by looking at all names in Sympl (there aren't that many), and seeing the ones that have "Wrapper" in the name. If we really need to be able to use tab to look at available wrappers, we could have an additional wrapper namespace so people could go "sympl.wrappers.<tab>" to see the list. There should also be a documentation page for this purpose.

There's the additional issue right now that if people are used to using prognostic_version on CliMT objects and then go to use Prognostic objects from other sources (or ones they've created), then these methods won't work on those objects, leading to confusion.

Remembering what to import (or looking it up) to get something done is standard practice in Python.

Reply:
I'm not contesting the fact that you can always look up what needs to be imported -- I do it as well, and in some cases, it gets quite frustrating (especially given my memory!). I don't think I'm the only one, and the current API makes it easy for people like me to work with CliMT.

And this is not the only issue. You import some things, and then if you realise you want to use one of the Wrappers, you have to go and import that as well (if you don't use IDEs like me). It just adds to the mental load, and reducing such a load is one of my main objectives.

CliMT objects subclass sympl objects and have different names (ClimtPrognostic etc.,), so I don't see why a person would presume that Prognostic would provide the same functionality as ClimtPrognostic, for example.

With regards to multiple ways of doing things, if anyone reads CliMT's documentation, they will see only one way to create these wrappers, and I think that this way of creating Wrappers is the most obvious way.

Coming to the issue of the returning a private type, I agree that sympl must export these classes.  The eventual aim is to be able to use CliMT without actually needing to explicitly import sympl (like how xarray works with dask). So, CliMT won't export these classes, and if anyone really wants to see what they are, they can do so by explicitly importing sympl -- just like xarray's open_mfdataset returns a dask array.

Reply:
I think the best practice for sympl is "import sympl as sp", then you don't need to go back and import each wrapper individually. I think if you do that, the effort/mental load/amount of code needed/etc. are equivalent between the two options (wrapping by accessing an Object, or by accessing a method). In that case, the main difference between the two is that one results in a more complex API with redundancies, breaking a Python idiom/best practice.

And I think no matter which way we do it, "wrapping an object" is a non-model-specific goal that should be fully implemented in Sympl. If it is necessary to add these methods to a subclass of Prognostic in CliMT, then that's saying something is wrong/insufficient in Sympl.

Reply:
There is a coding reason why it might be better/necessary to use these methods. If we're using methods, it's possible to ensure the returned object is a subclass of the right component type. For example, a Prognostic could return a PiecewiseConstantPrognostic while a Diagnostic could return a PiecewiseConstantDiagnostic. Then one could use isinstance(object, Prognostic) once again, something we've avoided exactly because of wrappers.

Reply:
sympl exports way more symbols than does a ClimtPrognostic, and I don't think using one or the other is comparable in terms of ease of use/searching. Again, I don't see how what we are doing here is different from what is already present in xarray/dask -- the CliMT namespace won't export Wrapper classes.

I don't agree that this means that there is something wrong/insufficient in Sympl -- It just reflects two different (and IMO, equally valid) visions of how to interact with Sympl components/wrappers. If in the long run we see that using the CliMT methods is making it more confusing for users, then I don't have a problem removing them.

Reply:
that is true. I like strong typing, so I don't have a problem with such a scenario! So, do you propose implementing these methods at the Sympl level? We could refactor before the upcoming release.

Reply:
Yes, we should implement these methods in Sympl. I feel comfortable justifying the API choice with strong typing. For now leave them as they are, but I may change the names slightly when we migrate them to Sympl (I dislike the use of the word "version" in the name, which implies older/newer type versioning). I'll g-chat you about the names.